### PR TITLE
WIP: Generate Changelog automagically

### LINKED ISSUE #505: Manage the changelog automatically
Currently writing the changelog is pretty cumbersome, we should take take the approach of Gitlab to write changelog records in yml format and automatically generate the changelog out of that. Every PR author can write a new yml file into a specific folder so that we can generate the changelog automatically. You can see the records of Gitlab at https://gitlab.com/gitlab-org/gitlab-ce/tree/master/changelogs/unreleased.

## Format

I think the format used by Gitlab is already pretty fine, but I would also add a `kind` flag for grouping:

```
kind: enhancement # Possible values: breaking, feature, bugfixes, enhancement, misc
title: A short title that explains the change
pull_request: 256
issue: 234
author: tboerger
```

* `kind`: Group the changelog records by this flag as we currently got for `v1.0.0` release
* `title`: A possibly short title to explain the change
* `pull_request`: Optionally the number of the pull request used as a reference
* `issue`: Optionally the number of the meta issue for multiple pull requests
* `author`: The name of the author who contributed the change

## Directories

For entirely new features I would suggest to put the file into `changelog/unreleased`, if it's already clear which version will be the target for this change we should put it into `changelog/v1.1.0`, for the filename I would suggest the format `YYYYMMDD-short-title.yml`.
 
## Scripting

To generate the changelog I would like to write a Go script to `scripts/` that generates the `CHANGELOG.md` file out of the yml files. We can even integrate this into the CI process and add a commit which includes `[skip ci]` in the commit message to update this on every merge on master and the release branches.

### PR DESCRIPTION
**Pull request by bkcsoft**
## What it does:

Automagically generate changelog on `make release` 🎉 

### TODO:
- [x] Generate changelog
- [x] Append historical changelog (kept in `changelog/HISTORICAL.md`
- [x] Move unreleased-section in current changelog to yaml-files
- [x] Updated the Makefile
- [ ] Make it efficient 😂 
- [ ] Document it :trollface: 

### Format: (stolen from #505)
```
kind: enhancement # Mandatory! Possible values: breaking, feature, bugfixes, enhancement, misc
title: A short title that explains the change # Mandatory
pull_request: 256 # Mandatory. Number to indicate which PR this was made in.
issue: 234 # Optional. Issue number that got closed by this.
author: tboerger # Optional. The author/authors of the PR
```

Current implementation is _extremely_ write-intensive... That should be fixed 🙄 

_Please_ be brutal when reviewing this, since **it needs to work** 100% flawless

I assigned all Owners to review, since _we_ deal with this pain. Also assigned Team maintainers since I want more eyes on this code 😆 

Closes #505

### CODE REVIEW

**Comment by bkcsoft** [comment]
@strk Since I moved your changelog to a file, do you want your full name in there or nickname? (or both)

File: scripts/generate-changelog.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+	{{ if .Breaking }}
+	* BREAKING
+	{{ range .Breaking }}
+	  * {{ }}
+	`
+)
+
+var (
+	version    = flag.String("version", "", "Version for changelog")
```
Comment (tboerger) [inline scripts/generate-changelog.go]: The definition into main to avoid conflicts?

File: scripts/generate-changelog.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+	outputFile = flag.String("outputFile", "CHANGELOG2.md", "File to which the changelog is written")
+	force      = flag.Bool("force", false, "Force update of changelog even if version already exists. DON'T USE THIS!")
+)
+
+func init() {
+	flag.Parse()
```
Comment (tboerger) [inline scripts/generate-changelog.go]: Within the other scripts I have moved that into main as well

File: scripts/generate-changelog.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+	}
+
+	log.Printf("Generating new changelog\n")
+	cl := newChangelog()
+	if err := filepath.Walk(changelogDir, cl.Populate(unrelDir)); err != nil {
+		fmt.Printf("Error: generating changelog: %v\n", err)
+		os.Exit(1)
+	}
+	if err := cl.Write(*outputFile); err != nil {
+		fmt.Printf("Error: writing changelog: %v\n", err)
```
Comment (tboerger) [inline scripts/generate-changelog.go]: Exit?

File: scripts/generate-changelog.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+const (
+	changelogDir = "changelog"
+	unrelDir     = changelogDir + "/unreleased"
+	relDir       = changelogDir + "/v"
+	relDirFmt    = relDir + "%s"
+
+	prUrlFormat  = "https://github.com/go-gitea/gitea/pull/%d"
+	tagUrlFormat = "https://github.com/go-gitea/gitea/releases/tags/v%s"
+
+	cleTmpl = `
```
Comment (tboerger) [inline scripts/generate-changelog.go]: Incomplete or unused?

**Review by tboerger — CHANGES_REQUESTED** [review]
Maybe generally use log instead of fmt?

File: changelog/HISTORICAL.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
+# Changelog
+
+## Unreleased
```
Comment (tboerger) [inline changelog/HISTORICAL.md]: Shouldn't be part of the historical file

File: scripts/generate-changelog.go (hunk 4 above)
Comment (bkcsoft) [reply scripts/generate-changelog.go]: Unused

File: Makefile [hunk 6]
```diff
@@ -101,7 +101,17 @@ docker:
...
 .PHONY: release
-release: release-dirs release-windows release-linux release-darwin release-copy release-check
+release: release-changelog release-dirs release-windows release-linux release-darwin release-copy release-check
+
+.PHONY: release-changelog
+release-changelog: CHANGELOG.md
+
+.PHONY: CHANGELOG.md
+CHANGELOG.md: changelog/unreleased/*.md changelog/HISTORICAL.md
+	cat changelog/HISTORICAL.md >> CHANGELOG.md
```
Comment (strk) [inline Makefile]: This sounds brutal. A Makefile target that always unconditionally appends to a file ?
I see that file is created by the unreleased/*.md target, but it isn't at all intutitive.

**Comment by lunny** [comment]
ChangeLog codes maybe should be in `modules/changelog` so that we can add a feature to gitea as `generate a changelog`. Then when we eat dogfood, we could use it and no need write twice.

**Comment by bkcsoft** [comment]
@lunny I thought `go run` didn't like modules. But now that I think about it, it makes a lot of sense to make it a package 😄

**Comment by bkcsoft** [comment]
@lunny Maybe even it's own repo?

**Comment by lunny** [comment]
@bkcsoft I just give an advice to reuse the functions. But I cannot give you some ideas about how to do that.

File: Makefile (hunk 6 above)
Comment (bkcsoft) [reply Makefile]: I've updated it so it hopefully makes more sense :)

**Comment by bkcsoft** [comment]
@tboerger I've updated according to your request :)

**Comment by bkcsoft** [comment]
@lunny @tboerger I made it into it's own repo. You can find it here: https://gitlab.com/bkc/changelogger

**Comment by lunny** [comment]
@bkcsoft great!

**Comment by lunny** [comment]
build failed.

**Comment by bkc-soft** [comment]
Gitea is stupid

**Comment by bkcsoft** [comment]
I'm closing this in favour of the actual package.

**Pull request closed without merging**